lib/offline/download_manager.js

```diff
@@ -168,6 +168,60 @@ shaka.offline.DownloadManager = class {
     return newPromise;
   }
 
+  /**
+   * Add a request to be downloaded as part of a group.
+   *
+   * @param {number} groupId
+   *    The group to add this segment to. If the group does not exist, a new
+   *    group will be created.
+   * @param {!BufferSource} queueData
+   * @param {number} estimateId
+   * @param {boolean} isInitSegment
+   * @param {function(BufferSource):!Promise} onDownloaded
+   *   The callback for when this request has been downloaded. Downloading for
+   *   |group| will pause until the promise returned by |onDownloaded| resolves.
+   * @return {!Promise} Resolved when this request is complete.
+   */
+  queueData(groupId, queueData, estimateId, isInitSegment, onDownloaded) {
+    this.destroyer_.ensureNotDestroyed();
+
+    const group = this.groups_.get(groupId) || Promise.resolve();
+
+    // Add another download to the group.
+    const newPromise = group.then(() => {
+      // Make sure we stop downloading if we have been destroyed.
+      if (this.destroyer_.destroyed()) {
+        throw new shaka.util.Error(
+            shaka.util.Error.Severity.CRITICAL,
+            shaka.util.Error.Category.STORAGE,
+            shaka.util.Error.Code.OPERATION_ABORTED);
+      }
+
+      // Update initData
+      if (isInitSegment) {
+        const segmentBytes = shaka.util.BufferUtils.toUint8(queueData);
+        const pssh = new shaka.util.Pssh(segmentBytes);
+        for (const key in pssh.data) {
+          const index = Number(key);
+          const data = pssh.data[index];
+          const systemId = pssh.systemIds[index];
+          this.onInitData_(data, systemId);
+        }
+      }
+
+      // Update all our internal stats.
+      this.estimator_.close(estimateId, queueData.byteLength);
```

Comment:
Should we be updating the estimator with this? I'm not clear from reading this if these datas we store are added to the estimated total. If not, calling `close` here would lead to the estimator saying the download is further along than it really is.

Reply:
This method is actually a copy of the method above, but what changes is that instead of making a request, the data is already available.

Reply:
Ah, so the size of the already-downloaded data is presumably part of the original size estimate that was opened? That makes sense.

Reply:
yes!